mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -288,41 +290,49 @@ public MessageMetaData add(Mailbox mailbox, MailboxMessage message) throws Mailb
 
         int retryCount = 0;
-
         while (retryCount < maxRetries && !globalResult.getFailed().isEmpty()) {
             retryCount++;
-            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId,
-                FluentFutureStream.of(
-                    globalResult.getFailed().stream()
-                        .map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
-                    .flatMap(OptionalConverter::toStream)
-                    .completableFuture().join(),
-                flagsUpdateCalculator);
+            Stream<ComposedMessageIdWithMetaData> idsFailed = FluentFutureStream.of(
+                globalResult.getFailed().stream()
+                    .map(uid -> messageIdDAO.retrieve(mailboxId, uid)))
+                .flatMap(OptionalConverter::toStream)
+                .join();
+
+            FlagsUpdateStageResult stageResult = runUpdateStage(mailboxId, idsFailed, flagsUpdateCalculator);
 
             globalResult = globalResult.keepSuccess().merge(stageResult);
         }
 
-        LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
+        if (!globalResult.getFailed().isEmpty()) {
+            LOGGER.error("Can not update following UIDs {} for mailbox {}", globalResult.getFailed(), mailboxId.asUuid());
+        }
 
         return globalResult.getSucceeded();
     }
 
     private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator) {
-        Long newModSeq = modSeqProvider.nextModSeq(mailboxId).join().orElseThrow(() -> new RuntimeException("ModSeq generation failed"));
+        Long newModSeq = modSeqProvider.nextModSeq(mailboxId).join().orElseThrow(() -> new RuntimeException("ModSeq generation failed for mailbox " + mailboxId.asUuid()));
 
-        FlagsUpdateStageResult result = toBeUpdated
-            .map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator,
-                newModSeq,
-                oldMetadata))
-            .reduce(FlagsUpdateStageResult::merge)
-            .orElse(none());
+        return runUpdateStage(mailboxId, toBeUpdated, flagsUpdateCalculator, newModSeq);
+    }
 
-        result.getSucceeded().stream()
-            .map((UpdatedFlags updatedFlags) -> indexTableHandler.updateIndexOnFlagsUpdate(mailboxId, updatedFlags)
-                .thenApply(voidValue -> updatedFlags))
-            .forEach(CompletableFuture::join);
+    private FlagsUpdateStageResult runUpdateStage(CassandraId mailboxId, Stream<ComposedMessageIdWithMetaData> toBeUpdated, FlagsUpdateCalculator flagsUpdateCalculator, Long newModSeq) {
+        return reduceResult(toBeUpdated.collect(JamesCollectors.chunker(UPDATE_FLAGS_BATCH_SIZE))
+            .values().stream()
+            .map(uidChunk -> FluentFutureStream.of(
+                uidChunk.stream().map(oldMetadata -> tryFlagsUpdate(flagsUpdateCalculator, newModSeq, oldMetadata)))
```

Comment:
I find this chunk of code difficult to read because it mixes very technical stuff with what is actually important. Could you try to make the code simpler ?